### PR TITLE
add terminal help widget

### PR DESCRIPTION
**Pull request by meganrogge**
Screenshot 2023-01-19 at 4 33 01 PM

### CODE REVIEW

**Comment by meganrogge** [comment]
unsure if we should mention navigation mode as it sometimes does not work

File: src/vs/workbench/contrib/terminal/browser/media/terminal.css [hunk 1]
```diff
@@ -543,3 +543,18 @@
...
+	z-index: 35;
+	position: absolute;
+	left: 0;
+	bottom: 0;
+	right: 0;
+	top: 0;
+	margin: auto;
+	text-align: center;
+	background-color: var(--vscode-menu-background);
+	width:50%;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/media/terminal.css:556]:
```suggestion
	width: 50%;
```

Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/media/terminal.css:556]: min-width too?

File: src/vs/workbench/contrib/terminal/browser/terminal.ts [hunk 2]
```diff
@@ -154,19 +154,19 @@ export interface ITerminalService extends ITerminalInstanceHost {
 	 * Creates a terminal.
 	 * @param options The options to create the terminal with, when not specified the default
 	 * profile will be used at the default target.
-	 */
+	*/
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminal.ts:157]: Let's undo these changes

File: src/vs/workbench/contrib/terminal/browser/terminalActions.ts [hunk 3]
```diff
@@ -176,6 +176,29 @@ export function registerTerminalActions() {
...
+		constructor() {
+			super({
+				id: TerminalCommandId.ShowTerminalAccessibilityHelp,
+				title: { value: localize('workbench.action.terminal.showAccessibilityHelp', "Show Terminal Accessibility Help"), original: 'Show Terminal Accessibility Help' },
+				f1: true,
+				category,
+				precondition: ContextKeyExpr.and(TerminalContextKeys.processSupported),
+				keybinding: {
+					primary: KeyMod.Alt | KeyCode.F1,
+					weight: KeybindingWeight.EditorContrib,
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalActions.ts:189]:
```suggestion
					weight: KeybindingWeight.WorkbenchContrib,
```

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 4]
```diff
@@ -1971,7 +1983,7 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
 
 	private _setAriaLabel(xterm: XTermTerminal | undefined, terminalId: number, title: string | undefined): void {
 		if (xterm && xterm.textarea) {
-			let label: string;
+			let label: string = this._accessibilityService.isScreenReaderOptimized() ? nls.localize('terminalTextBoxAriaLabelNumberAndTitleAccessibility', "Press alt+f1 for terminal accessibility help") : '';
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:1986]: This will get overwritten by setting `label` below, we want to combine these messages, with the previous message first. Eg. `Terminal {0}, {1}. Press alt+f1 for terminal accessibility help`

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 5]
```diff
@@ -2642,3 +2654,73 @@ export function parseExitResult(
...
+		if (this._hasShellIntegration) {
+			content += nls.localize('shellIntegration', "The terminal has a feature called shell integration which offers an enhanced experience. \n\nIt provides some useful commands for screen readers such as  ");
+			content += strings.format(nls.localize('runRecentCommand', 'Run Recent Command ({0})'), runRecentKbLabel);
+			content += ' and ';
+			content += strings.format(nls.localize('goToRecentDirectory', 'Go to Recent Directory ({0})'), goToRecentKbLabel);
+			content += '\n\n';
+		}
+		content += strings.format(nls.localize('detectedLink', 'The Open Detected Link ({0}) command enables screen readers to easily open links found in the terminal'), openDetectedLinkLabel);
+		content += '\n\n';
+		content += nls.localize('minContrastRatio', "The setting terminal.integrated.minimumContrastRatio takes a number between 1 and 21 and will adjust the text color luminance to meet this\n\n");
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:2721]: Let's not mention this as it's for sighted only, so it's likely not applicable for screen readers.

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 6]
```diff
@@ -2642,3 +2654,73 @@ export function parseExitResult(
...
+		content += nls.localize('dismiss', "You can dismiss this by pressing Escape or focusing elsewhere\n\n");
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:2722]: Should we point to the docs towards the end?

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 7]
```diff
@@ -2642,3 +2654,73 @@ export function parseExitResult(
...
+	private _buildContent(): void {
+		const runRecentKbLabel = this._keybindingService.lookupKeybinding('workbench.action.terminal.runRecentCommand')?.getAriaLabel();
+		const goToRecentKbLabel = this._keybindingService.lookupKeybinding('workbench.action.terminal.goToRecentDirectory')?.getAriaLabel();
+		const enterAccessibilityModeLabel = this._keybindingService.lookupKeybinding('workbench.action.terminal.enterAccessibilityMode')?.getAriaLabel();
+		const openDetectedLinkLabel = this._keybindingService.lookupKeybinding('workbench.action.terminal.openDetectedLink')?.getAriaLabel();
+		let content = nls.localize('introMsg', "Welcome to Terminal Accessibility Help\n\n");
+		content += strings.format(nls.localize('enterAccessibilityMode', 'The Enter Accessibility Mode ({0}) command enables screen readers to read the terminal buffer'), enterAccessibilityModeLabel);
+		content += '\n\n';
+		if (this._hasShellIntegration) {
+			content += nls.localize('shellIntegration', "The terminal has a feature called shell integration which offers an enhanced experience. \n\nIt provides some useful commands for screen readers such as  ");
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:2713]: A more elegant and better with memory pattern here is:

```ts
const content: string[] = [];
content.push(...);
content.push(...);
content.join('\n\n');
```

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 8]
```diff
@@ -2642,3 +2654,73 @@ export function parseExitResult(
...
+			content += strings.format(nls.localize('runRecentCommand', 'Run Recent Command ({0})'), runRecentKbLabel);
+			content += ' and ';
+			content += strings.format(nls.localize('goToRecentDirectory', 'Go to Recent Directory ({0})'), goToRecentKbLabel);
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:2716]: They keybindings could be undefined, not sure if there are better ways to do this but you could create a helper that returns the text as either `{0} ({1})` or `{0}`

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts (hunk 7 above)
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:2713]: Double space at the end

Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:2713]: The shell integration introduction should be a single paragraph

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts (hunk 8 above)
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:2716]: Should these be list items?

File: src/vs/workbench/contrib/terminal/browser/media/terminal.css [hunk 9]
```diff
@@ -543,3 +543,18 @@
...
 
+.monaco-workbench .terminal-accessibility-help {
+	z-index: 35;
+	position: absolute;
+	left: 0;
+	bottom: 0;
+	right: 0;
+	top: 0;
+	margin: auto;
+	text-align: center;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/media/terminal.css:554]: Left aligned would probably look better, especially if we use list items.

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 10]
```diff
@@ -2642,3 +2654,73 @@ export function parseExitResult(
...
+	}
+
+	public hide(): void {
+		if (!this._container) {
+			return;
+		}
+		this._domNode.setDisplay('none');
+		this._contentDomNode.setDisplay('none');
+		this._domNode.setAttribute('aria-hidden', 'true');
+		this._contentDomNode.domNode.tabIndex = -1;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:2691]: You can set it to 0 once and not on hide/show because it doesn't have an effect when display is none

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 11]
```diff
@@ -2642,3 +2654,73 @@ export function parseExitResult(
...
+		this._contentDomNode.domNode.focus();
+	}
+
+	private _buildContent(): void {
+		const runRecentKbLabel = this._keybindingService.lookupKeybinding('workbench.action.terminal.runRecentCommand')?.getAriaLabel();
+		const goToRecentKbLabel = this._keybindingService.lookupKeybinding('workbench.action.terminal.goToRecentDirectory')?.getAriaLabel();
+		const enterAccessibilityModeLabel = this._keybindingService.lookupKeybinding('workbench.action.terminal.enterAccessibilityMode')?.getAriaLabel();
+		const openDetectedLinkLabel = this._keybindingService.lookupKeybinding('workbench.action.terminal.openDetectedLink')?.getAriaLabel();
+		let content = nls.localize('introMsg', "Welcome to Terminal Accessibility Help\n\n");
+		content += strings.format(nls.localize('enterAccessibilityMode', 'The Enter Accessibility Mode ({0}) command enables screen readers to read the terminal buffer'), enterAccessibilityModeLabel);
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:2710]: We should use full stops

**Review by Tyriar — DISMISSED** [review]

File: src/vs/workbench/contrib/terminal/browser/terminal.ts [hunk 12]
```diff
@@ -154,19 +154,19 @@ export interface ITerminalService extends ITerminalInstanceHost {
...
 	createTerminal(options?: ICreateTerminalOptions): Promise<ITerminalInstance>;
 
 	/**
 	 * Creates a raw terminal instance, this should not be used outside of the terminal part.
-	 */
+	*/
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/terminal.ts:162]:
```suggestion
	 */
```

File: src/vs/workbench/contrib/terminal/browser/terminal.ts [hunk 13]
```diff
@@ -154,19 +154,19 @@ export interface ITerminalService extends ITerminalInstanceHost {
...
 	getInstanceFromId(terminalId: number): ITerminalInstance | undefined;
 	getInstanceFromIndex(terminalIndex: number): ITerminalInstance;
 
 	/**
 	 * An owner of terminals might be created after reconnection has occurred,
 	 * so store them to be requested/adopted later
-	 */
+	*/
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/terminal.ts:169]:
```suggestion
	 */
```

File: src/vs/workbench/contrib/terminal/browser/terminal.ts [hunk 14]
```diff
@@ -178,13 +178,13 @@ export interface ITerminalService extends ITerminalInstanceHost {
 	 * Perform an action with the active terminal instance, if the terminal does
 	 * not exist the callback will not be called.
 	 * @param callback The callback that fires with the active terminal
-	 */
+	*/
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/terminal.ts:181]:
```suggestion
	 */
```

File: src/vs/workbench/contrib/terminal/browser/terminal.ts [hunk 15]
```diff
@@ -178,13 +178,13 @@ export interface ITerminalService extends ITerminalInstanceHost {
...
 	doWithActiveInstance<T>(callback: (terminal: ITerminalInstance) => T): T | void;
 
 	/**
 	 * Fire the onActiveTabChanged event, this will trigger the terminal dropdown to be updated,
 	 * among other things.
-	 */
+	*/
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/terminal.ts:187]:
```suggestion
	 */
```

File: src/vs/workbench/contrib/terminal/browser/terminal.ts [hunk 16]
```diff
@@ -194,7 +194,7 @@ export interface ITerminalService extends ITerminalInstanceHost {
 	 * @param linkProvider When registered, the link provider is asked whenever a cell is hovered
 	 * for links at that position. This lets the terminal know all links at a given area and also
 	 * labels for what these links are going to do.
-	 */
+	*/
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/terminal.ts:197]:
```suggestion
	 */
```

File: src/vs/workbench/contrib/terminal/browser/terminal.ts [hunk 17]
```diff
@@ -154,12 +154,12 @@ export interface ITerminalService extends ITerminalInstanceHost {
 	 * Creates a terminal.
 	 * @param options The options to create the terminal with, when not specified the default
 	 * profile will be used at the default target.
-	 */
+	*/
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/terminal.ts:157]:
```suggestion
	 */
```

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts (hunk 8 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:2716]: I thought you said we don't want a list of commands?

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:2717]: They won't be undefined unless the user has changed them from the default - in which case I think it's important for the user to see that they are undefined

**Review by gregvanl — APPROVED** [review]
